### DISCUSSION TITLE
Stack overflow when running a recursive programming language grammar

### DISCUSSION THREAD

**Discussion by TheOnlyTails** [pest-parser/pest#1028] [Q&A]
I'm trying to write a grammar for a custom general-purpose programming language using pest, and while I've managed to write the grammar, running it always results in a `fatal runtime error: stack overflow`.
[Here](https://pest.rs/?g=N4Ig5gTghgtjURALhDA9gEwK4BsCmABALwED6wBAygPICSBAfgRngMY4JQAuAlmgHYAqRgQCidAgF8AOv1kB6RQQCCzNh2i8BBbgS4ALQlzQAHAvgBueHATQAzHQXTZ8slu05b%2BBHz5IUeGBM0CC5SdxsAHwILHgBnHgAjHhweLgBPcPUpWViE5NSMrPZiAgo8pJS09IB%2BEQAKDJM8UihUqDjiqII4rggsVjCIgmi8fiwYLpHzPCHs6LssflYp6J5%2BLjwIOyhWFuG1oJxSdc3t3cJDk2PTrZ296f5YPDiTC66AShy5fkV5FRi8UqhXSMysNnsOm8aTwMAAdLkgQVqqUKNIQCYsIlUqx0dN0bcIE8cHjouiTBAeBZuHg8TIfn8VNCgiEuGoPJo%2BPwADQEVgCLhQdZxRxxDhxfQAWjieDemjwGHM8TZkOcuEIPBYGx4dh4WziCiUUH4itMXjaOHSvKgSt6tgcmrGvF1%2Br0aAIgWCoQIdggaBgBAA7ml9Os9PpdGr8Aj%2BJ7WVNUQQCSzQnimI7tS6IA10fI0x6tc69RAPsImLn0XUmPV0cGDPn0TWQCI46EToWcyBufnW2EM1xSyJ0d2QFWkyAPuiPnV6b3207E-2dcXOx0e4Xl1tp99Df9lN42jwOvaoWgDFs9OlmjGmi1Dx0pv5x7f107N9mmGAxlseCs5bA4jHdEiHzW8sgcelZF6fpBgTXx4L8Mpx2ggYuFfTMV0-b9KT-TgYEAocQGAfMUNgl0cAwEVqyeGAXjePZSGxNBWAAa2mT1jlomBEgvK4bg2O4LkHcsQBkZt6VIsJyMohD4KfepJNIaTCJHYSCAUvpUKUvUKJUyttwkzSyJ0xVZN8J8KmRDIx0U6TSGcd8ywLN8s0IpBQKvfY8AcatgPzPAAA8KQMqCjKkkz7Mwd9E3RGAsDQ8TZFkMYJjgsyYpAFKYHQosLyw-gf1w6B8MI4jmyYLLSGpSljS4Kj1Jouj3iY1j2KOezYR47M%2BJOATzj2NT0TE75Kuqo8NhFWT5NGhBxrZETVKc%2BoZpqjY9NHEKCvGSYxtq9KLKRKprJbDdXN8oj-O2qrZtq7TrBkkSxM2la5ruiiRSfWyTM7EcRC%2B%2B7Bq7Ssd34fA5hKOSkPRMH8zByKMEck6XMw9T0Xc8rL2acDtxEkCMcCilvjhhzXKfWL4rpJL%2BEWZYEzJkAadxDGlzOggvwKnDSH-fCgJAJsGkZrm8PWtT6kF7n1pqHHx0nDGm3RkQwJYOxpfRSUAD5-KC7N6XF4X6bhQ3gaYBJ%2BDAfAha4M5vBEhWmCV7zOzxkQCZLGcqcJe4vIhhCDvyI7akIz2Lhy6L8sKoXioIkSypbPqvY67itiWxrXmanBmLYnrg4Gpyhsp2N4-eLiut9pDLIDscc5aGEA1nIuGNrsyn2rpSllYdiG5aXbsUuD0u9IMD7xFeuzgTxnm6hhn29D1n2Yj7no-Hfnqz14qRaWtfYEl1WJ3zeWPKx5XpgPjGHZV3mNa14Lvlbnv8HMqeYeZ06UdPxXPPAk%2B0cPryL8I52FVtZfFHoJBiQ92gfSni%2BF%2ByM8ps2wr%2BSOAExz1HfmSEAeMvj20-srTasgOK9THhcRMFcQS8w4rPFG89OYIDABMJ0S90Sx3TO1EuydCLDVnO1QKmx%2BAJG0H7YE1QKFHHzDQpBi8q6v3gRIlYdCGETV5nYEIv8v4xx7GwzqHDHoF0IcKQUyxCAZUoRjZhmjriJy6nnUSeitFJ2zI-coh1yEtnak3bhlim5lwoHDA4Pp26rExneSBUxIL8FTvRb2NhSEuJEYRSJIdYEYVkYgoqKDSo9gHuwiANiuGyESQxFqbF6aFNpGYi6zNsnaNyZwguoUaS0TWk%2BOGvQYBsmiBAWYWAiSkFdtMfp9J%2Bk%2BIIBSbyPAAqkFMGOV2rRjABiYMEXoupJnTIaMkJ4EBMimBEGMlZUyTAzO1nM-0uy0DLImQcqWwghnHO4Kckg5AlRnDaH07W0w0CJAAFZsDCIULYbRpiQDQFgZoGA3mE2iOwc5PSWj-OgN0fk2pxgtH6dERIXSoAsQhd1Ag0AzZwphAi6YBh4g4umNcC4%2Bg0AUS2OS9FGdWL0p9CEZlgZQwWzRR6OwzL4BcFYPoZlLMVzhPha8-pjyKB2AztwUgYruinDlUSwF6K0A0rwMaJVLzEURggFqgF3RoLrDAPq4l0RUi9FNSqvQoKLbyumPAEwVqbDhMVfayGAABCgGyECZFOB8wYrz-XREMAFKA7hAhBrWtECN8B%2BJsnpD6rZRCpoEC9QQAAeuiAADIkfeyhKAAGFaC0FIAAIVoAAOVIAAEVoAAcVoAAFXxCAUgU4nIFuLaWit1a62NpbdWLtJby1VtrQ25trb20ThubIZigp42T3TVmkA2a0D5qLSO6gham3joHVOjtIhh2lu3bu-tk6h2bpPTuvdk6MHTsHPSUN4a2CRsXSQZdOaAobu7aQAAEqIAAGrelt97D1MGPf%2BoDIGGiQYA8B89oHxwPtnSaV9caU2eooPUSDiGD0zqPVemDl7f14bAwR8J0q0CyvdR%2Bigsao3zXHHYUk7I30ppEgiDGDHF0xvQ4xzsXGRA8aIdLKjsqCYCAXLrGVYRJNbTZHRzN6JymdklKx9EABqKcY4RP%2BvCUoaQRnpC8iMzyJM0gICmekGhJ4cRWA8B4KQGUABHLAYwHj02M0ZuW6JvMaZAHIZsGCIABYSiAw0FmsDKDLYWmtsgli-kwKi%2BzUBmgZW81gH9I74MwfIx8YAABGXkAA2SQ3xDPSAACTEULmcYIHAvB9NS%2BlrzxmasF0qwAcms35-gbmzwpdYGlwgbWjNdYC-5nznX-g1ZUWgeoHxQEQAazScFEqCA1fHB1-G7zdGJQZP8LrUAJvLF1c60oU8JtVI5qQAVCBCLXdvhEul93HGlHTR0BzTnXPuaMdMRL-IWDNeG%2Bl6IA3Ngg5GyfAAhE2J7GD-PBYIJW0QAB1AAMlW0Q2CVCVoAJoRcO%2BOKAZmjNYEKwAJmUA4ObaqxKhUpGbC7j8tt%2BemzdulRqzaEWMyAPRL29Xc7AKm9N9R1i3eF3d3Vj6CGC%2Bc30Y10uHskCQl9xzzm8BuY833QHyWodg4IBDob0O1h9VW01l4oO%2B6EgtwqZl9Q4fs788j53Pnpio8x9j3HygCdE9kIkNV%2BBNX2qU%2BiTSqmME7BwDKepsYZHZiU-UQDtAa2kEoE25QAAlJD6IH0iBT2nwt1BK1NqrQAVVEKhyLABiAA2sVgglPeQAGZeQABYAC6sgLV-OVTYem9f979IWuB9SI-xwjilrzbvB2lA1-qI347PXxyGBwBnYceh%2Bh4EnBsW1hLtUZQX8Pvbk%2Bx-LVP8OfSvNZYg3n8AZfSACBL8SF1p-zeCBddYG-ggLf6SOou0HxYXUgAKdC2Q3gaFAI2HANHw2l5nyX4CgL6FBEQgoAnx-l20JgM3%2BAAAV9RtAKBGpSh0RK1nhN8oAvxShKdG9DY4Quk7R6RPkflYJQ8kJhV4FzE5YmDfk%2BloDQRYDRZuDYIwD%2BCz84C6kDshC5M%2BCnFxxaDr5cV2CPxxxAECAuUlC79-gAAeFtJ-cTLgdWeoKAJ-JtXkMtEwj4EwggDWRwTTAgPNZYDOOIWFQApCORZBHmQiFedSLeBZMQ0WXwneG-feDAj%2BI%2BbyXeK%2BTAnWKmJQQrQ2D-eIuEXkBIlI5I5-Q2IgD-TIynWQfFL8VwtA7WXmI2aIuoUYYowiTIhQkGPZS5HZSVAgAqMAGkR4M8b4Zo1ox%2BKedTA7fgdo1NKeGHOPJZLgfZHZSGJ5dYVgLpJpDpdkWYhcKFDoeYsCAVNgLOJos8QeT%2BdY1qaIHJVoVgPYOIEUM3FgSZXYE4s4gJWmYbdfZ7GY2EBcMuKeTTbTA7dwRYtaV4tEEASUXo74YbO0dKRCKeNcM%2BXBR2ekNYwwVqSGKeeINRY%2BekfosIWEjYy7P4mHJEyE8IiCWQQ4q4l4SaVXP4oTdMBPZ7C4o4647ov4uvGokSWfTQ241gLQ04Qw9ENfDfLsew0oFvXfQWe4mJWQ9whROYphPmfeEU1oCAehOYiApgWUiUp0YIg7FU%2BUxRRTJCcXKk3GHTKo0oo5LAqmJNbZMwSVWQeCa4LAToUwa0h1dYO0g5R06IXgWie0kwN05gKkR0V07wB1TAXAF0h0wM6IYIQMOlMMnwdFdYVoE0AM2M%2Bw%2BM1lGM6YDZUgAKNM708MmYOwMICUHUMIdMzpHgMAfQQs0MAspMgZNzNoL0n0tEvpesmPWs81EkweCMfgdsmYU4rszVLXLABs3syADVM4Acns0stmTFCcgwQc1sxsvM1M3M5M5slcn0wPGlBM8Facrc44HMn0joBIMAKc1cilHAF0488ss8n0mAZ0zoa8083sj0l4VoU4m83shGWIYHJ828vMqMK8j8586cyMulP83s-cnc98k8-85MqC1lCCvc%2BM40cFJC88uMqcvVdCzc%2BM7M7C4CuC6YfAGsosmsnCvMykCsqs4smCz8sM8JW0r00EyeP4j474e88YZiliyGP4wEmEwIN8iYni1A8cQQAub8-04Sni%2BmPMA7QC7imSqeAAUgLjAr1WkqUr%2BMEHEoO0zNQoOREoygADIC5MycyjKMpIgzK8KLKRL6YV0iZvIaKazNKRl0QtCtCC4qLKznNqySyzB9op51ZNYDshyRy3Kgq-iiBnZUTtjwq2zIrZD0QYdYqe9Oz5yzytLxwvKDt8B%2BzMqWzhzErArRKPK0qzZZy6VCqkr6ZQrvgxyaQ9VCqErmK6qKqVzLLeLxx1gC51ysqjL6YcSgtvgoKDKkrur0RjLTK9Kg8pkNLSqWL6ZIhrKDtZB0LBqp4KqmK6KQLFrJqQBNMKrOKgLYLDLU0%2BKKrXzHzCLzr3KQBBAKrJLfzbq3LZLjrgzLybqzrNL6ZlLtq0AoyCKfr9qyqHrHqzK5qDKNrkqQBpqKqELgb6L9rlrIgEaULEyYbYbjL0asLdqBqDq0abKez8L8a7qwaM0KqSKXKwgYb6ZPKKqfKaayaGjgr1YKrZAkVeAUVyV6YubnTVNzoPUqEtx3Z%2BAMUNVsVhkwaJasUQiQBhbklco3ZaiOA9hqVaU9UNs-jp1vhSVOhpb7r9a49ilebdSNCRJFbd5gDehGknQlp5VmVgVQV7cuVoVnCukLsoUBQBbmVZapb3lOljQCj7V3TQwDbA7Rk1a8ANaWAtbgF4C48ukuAekeztbxxk7U75bFakYUkVagEKQxbIswZHASBG8S64o2QF8oBeRSAvgSBj8QBSc%2BTCtd9WkuB2kMpn4RBiYooswbJjULY3grYtgbZUZkBkSIjeZVDBkqYVF47CYwb57s6Rb87l58xh7rYg4RqC7lCpx8xTbZ7%2BB2UUhUV3l6YT7XBfMFbV7d5vD0C94MZD73lsDuV1JCtiASBKcvhw9t9hjmxrAZRmMGy8B5D%2BdYweV06CQWMKl77L9H6XZT8mxAHBa1CE6QY%2BUBUza-jMH9AGxpTojCJb8NEMZcG5SSp9sOLuAsGEASonwyHaGfp8wGGIAYBAZfp-9qHBVGH6GuHyG7sBAEYvBCIojEHTTEC%2BHaGBGTQ0guRRLN7R7OwdQFD8Ffh-hF8CA7Dv7ZBnawVsH17CGRJb9GLuAt7WLxxVqB6zYh7THFHzpVq-pB6WgFGiRZd%2BEnHLYt6nx8jnHbGiQKU-Gez7k65ZAXGgn5lZDHawn2JCxPHFH3TP5omIzo7Y66Ukm8UXgwh0mpC4n-GDi0tcnvB3T99CniLlRSnohdH7donX6kiiA29n8AAOQ2NIogQrbNbNPI4O3xke-xkZHxwpk4TYGAEouEVQgZsJoZ2EaYCZwJqZvw9EUo74WZ3pnspuJ8N7L28cXoquDYLZtWY2H0WTC7V%2BiPXkZfUzEAHk9dPkyURvSnOEFvHvfvUpp8fcjVHs0Ovkc7L5qXL5g50cOOPvQ-DBHZkQfQk5j2WJ6Ji65yPO74SldWmlOO15qeXW%2BkegrJwJqKuQpZ1%2BvAiAOIAgpo54a0Sgsuxp74HJmF2Fi28cYA%2BoaluZ6SeqJsX6JgTFwp6fGZzJwp3HfOSQ75HgyZllxMJl1Zt6RUc6dl2wIV2CEVkyNx8V62SV0SulpsVQsJ1R%2B-T-E7J-JI85gALx-xyJKyoadRpYOprwZYAIVfuiVJAIKbtd0lgO5coc4YteZe%2BhaReYn3fkWUCe1f%2BHr2zVafucNm71BnKeicH0ZIxjCdQSv3jcDZsSn0IhZPpHn0Xxad1a60Ni62Nd5AJaJe8EINJZ0HJaaakF3xTuuB6a8ankboaGidgJ5ctXSZbfHpla1bsNdd3mMdiP%2BHiH7i4F5GHfr2F0jbAhjcRINDxL-hBiqfBRnb%2BO8M7f3rWoJQ5kkTwkmjBtypEHcIlmlfETSQ8PYcOfRHqvpCPeFm6vVdCJwXxOlg1cnv-hvbPdVImgygPfDloS1MVJPYxnFIA6dAvcBZEmvZ0c-dA5%2BPkjpb8gnDHHPhBhQ9haHgierDAnNMMqfbvHmTcYw9OR8UdJ8DAnlR9PI%2BVEo8-lrdcDzLAkdRo6xkZmY5aC6TsC2B1zY96k44gC6QwB46XdkBhM-ndRMQ2AC30ICzewC2FwC3eeNACwsDQE1ACwKisFCwOyo5BKtbjbCIXeZILjAjo%2BMVhqbaw6hJ8n8KWhQ77fVL1s-kdXMfRGtbfbcnc-ddkDAgnnuu8MaCs4dYC-xKCKIfwdEbw6-lE6xg464-%2Bxbn1IQW3fkVg7qjFrAnWD44E7BvRe88-iXdhu8Ls5ljj2w-WF9XJsaIy76hlEGDkei5rhq9%2BTkbJPHHYvCTRz-WbVEEoBwOUELVEFKCeU6%2B6-Tz64G%2B%2BGLwAFkpvRBS9JjsNcw5KGgYdPcsdUcj0-dBAvhogmx5BdKVv0RBBluIMtvCJjupxvguwQB1hMQuBkAQAS7-Uy70q2R9DSh4jKcW88BJR29XvvmVdc3-vhdiCQAdsatrnJAoAatCsxIGdQZZh7Cg8PnSgI9-ve9Sh697nW9I2K60tMeH8TsPvznX8qDznv8BT6QS73bYVSh6h8IwAn9hcvgbCKRTgcB%2BB6e4gwBd9RIgA#editor) is the grammar I'm using, together with the test file I'm running it on (though it fails on smaller examples too).

**Comment by tomtau** [comment]
have you tried stepping through it in a debugger: https://github.com/pest-parser/awesome-pest/blob/main/debugger.md ?

**Reply by TheOnlyTails** [reply]
From what I can see, it had a problem with the order of the `float_literal` alternatives which I fixed, but for some reason, when trying to parse an identifier as an `expr`, such as `a`, it goes into an infinite loop.
